File: python/interview/classic/ring_buffer/ring_buffer.py
"""定长环形缓冲区：队尾入、队头出，满/空抛 BufferError。"""

from __future__ import annotations

from typing import Generic, Optional, TypeVar

T = TypeVar("T")


class RingBuffer(Generic[T]):
    def __init__(self, capacity: int) -> None:
        if capacity < 1:
            raise ValueError("capacity must be >= 1")
        self._cap = capacity
        self._buf: list[Optional[T]] = [None] * capacity
        self._head = 0
        self._tail = 0
        self._size = 0

    def __len__(self) -> int:
        return self._size

    def is_empty(self) -> bool:
        return self._size == 0

    def is_full(self) -> bool:
        return self._size == self._cap

    def push(self, item: T) -> None:
        if self.is_full():
            raise BufferError("ring buffer full")
        self._buf[self._tail] = item
        self._tail = (self._tail + 1) % self._cap
        self._size += 1

    def pop(self) -> T:
        if self.is_empty():
            raise BufferError("ring buffer empty")
        item = self._buf[self._head]
        assert item is not None
        self._buf[self._head] = None
        self._head = (self._head + 1) % self._cap
        self._size -= 1
        return item


if __name__ == "__main__":
    rb: RingBuffer[int] = RingBuffer(3)
    rb.push(1)
    rb.push(2)
    assert rb.pop() == 1
    rb.push(3)
    rb.push(4)
    assert rb.pop() == 2
    assert rb.pop() == 3
    assert rb.pop() == 4
    assert rb.is_empty()
    try:
        rb.pop()
    except BufferError:
        pass
    else:
        raise AssertionError("expected BufferError")
    rb.push(10)
    assert rb.pop() == 10
    print("ring_buffer OK")

File: cpp/interview/classic/ring_buffer/ring_buffer.cpp
// 定长环形缓冲区（int 示例，可模板化）
#include <alg_std.hpp>
#include <cassert>
using namespace std;

class RingBuffer {
public:
    explicit RingBuffer(size_t cap) : cap_(cap), buf_(cap), head_(0), tail_(0), size_(0) {
        if (cap == 0) throw invalid_argument("cap>=1");
    }

    bool empty() const { return size_ == 0; }
    bool full() const { return size_ == cap_; }

    void push(int x) {
        if (full()) throw runtime_error("full");
        buf_[tail_] = x;
        tail_ = (tail_ + 1) % cap_;
        ++size_;
    }

    int pop() {
        if (empty()) throw runtime_error("empty");
        int v = buf_[head_];
        head_ = (head_ + 1) % cap_;
        --size_;
        return v;
    }

private:
    size_t cap_;
    vector<int> buf_;
    size_t head_, tail_, size_;
};

int main() {
    RingBuffer rb(3);
    rb.push(1);
    rb.push(2);
    assert(rb.pop() == 1);
    rb.push(3);
    rb.push(4);
    assert(rb.pop() == 2);
    assert(rb.pop() == 3);
    assert(rb.pop() == 4);
    assert(rb.empty());
    try {
        rb.pop();
    } catch (const runtime_error&) {
    } catch (...) {
        assert(false);
    }
    rb.push(10);
    assert(rb.pop() == 10);
    cout << "ring_buffer OK" << endl;
    return 0;
}
